go/marshal/decode.go

```diff
@@ -386,32 +438,45 @@ func interfaceDecoder(t reflect.Type) decoderFunc {
 	}
 
 	return func(v types.Value, rv reflect.Value) {
-		t := getGoTypeForNomsType(v.Type(), rv.Type(), v)
+		t := getGoTypeForNomsType(v.Type(), rv.Type(), v, nomsTags{})
 		i := reflect.New(t).Elem()
-		typeDecoder(t)(v, i)
+		typeDecoder(t, nomsTags{})(v, i)
 		rv.Set(i)
 	}
 }
 
-func getGoTypeForNomsType(nt *types.Type, rt reflect.Type, v types.Value) reflect.Type {
+func getGoTypeForNomsType(nt *types.Type, rt reflect.Type, v types.Value, tags nomsTags) reflect.Type {
+	listType := func() reflect.Type {
+		et := getGoTypeForNomsType(nt.Desc.(types.CompoundDesc).ElemTypes[0], rt, v, nomsTags{})
+		return reflect.SliceOf(et)
+	}
+
+	mapType := func() reflect.Type {
+		kt := getGoTypeForNomsType(nt.Desc.(types.CompoundDesc).ElemTypes[0], rt, v, nomsTags{})
+		vt := getGoTypeForNomsType(nt.Desc.(types.CompoundDesc).ElemTypes[1], rt, v, nomsTags{})
+		return reflect.MapOf(kt, vt)
+	}
+
 	switch nt.Kind() {
 	case types.BoolKind:
 		return reflect.TypeOf(false)
 	case types.NumberKind:
 		return reflect.TypeOf(float64(0))
 	case types.StringKind:
 		return reflect.TypeOf("")
-	case types.ListKind, types.SetKind:
-		et := getGoTypeForNomsType(nt.Desc.(types.CompoundDesc).ElemTypes[0], rt, v)
-		return reflect.SliceOf(et)
+	case types.ListKind:
+		return listType()
+	case types.SetKind:
+		if shouldSetDecodeAsMap(rt, tags) {
```

Comment:
This change looked right, but now I realise it isn't. Reverting (just "getGoTypeForNomsType").